lib/libc/cheri/strfcap.3

```diff
@@ -0,0 +1,173 @@
+.\" DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS
+.\" OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION)
+.\" HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
+.\" LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY
+.\" OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF
+.\" SUCH DAMAGE.
+.\"
+.Dd February 11, 2021
+.Dt STRFCAP 3
+.Os
+.Sh NAME
+.Nm strfcap
+.Nd format decoded CHERI capability
+.Sh LIBRARY
+.Lb libc
+.Sh SYNOPSIS
+.In cheri/cheric.h
+.Ft size_t
+.Fo strfcap
+.Fa "char * restrict buf"
+.Fa "size_t maxsize"
+.Fa "const char * restrict format"
+.Fa "uintcap_t cap"
+.Fc
+.Sh DESCRIPTION
+The
+.Fn strfcap
+function formats information from decoding
+.Fa cap
+into the buffer
+.Fa buf
+according to the string pointed to by
+.Fa format .
+.Pp
+The
+.Fa format
+string consists of zero or more conversion specifications and
+ordinary characters.
+All ordinary characters are copied directly into the buffer.
+A conversion specification consists of a percent sign
+.Dq Ql %
+and one other character.
+.Pp
+No more than
+.Fa maxsize
+characters will be placed into the array.
+If the total number of resulting characters, including the terminating
+NUL character, is not more than
+.Fa maxsize ,
+.Fn strftime
+returns the number of characters in the array, not counting the
+terminating NUL.
+Otherwise, zero is returned and the buffer contents are indeterminate.
+.Pp
+The conversion specifications are copied to the buffer after expansion
+as follows:-
+.Bl -tag -width "xxxx"
+.It Cm \&%a
+is replaced by address of the capability.
+.It Cm \&%A
+is replaced by a textual representation of capablity attributes
+enclosed in parentheses.
+Zero or more of the follow attributes are included:
+.Pp
+.Bl -tag -compact -width invalid
+.It invalid
+Capability's tag is clear.
+.It sentry
+Capability is a sealed entry.
+.It sealed
+Capability is sealed but not a sealed entry.
+.El
+.Pp
+If no attributes are present, the field is omitted.
+.It Cm \&%b
+is replaced by the base address of the capablity.
+.It Cm \&%c
+if
+.Fa cap
+is NULL-derived this is equivalent to
+.Ql 0x%a .
+If
+.Fa cap
+is not NULL-derived it is equivalent to
+.Ql 0x%a [%P,0x%b-0x%t]%? %A .
+This is equivalent to
+the
+.Ql %#p
+format of
+.Xr printf 3 .
+.It Cm \&%p
+is replaced by the raw permissions in hexadecimal format.
+.It Cm \&%P
+is replaced by an abbreviated textual representation of permissions
+listing zero or more of the following characters:
+.Pp
+.Bl -tag -compact -width W
+.It r
+LOAD permission
+.It w
+STORE permission
+.It x
+EXECUTE permission
+.It R
+LOAD_CAP permission
+.It W
+STORE_CAP permission
+.It E
+EXECUTIVE permission (Morello only)
+.El
+.It Cm \&%t
+is replaced by the top address of the capability.
+.It Cm \&%T
+is a pseudo operation causing the capability to be treated as tagged
+in future expansions and is replaced by the empty string.
```

Comment:
Oh, I see... is the intention that `gdb` (or `truss`) use something like `strfcap(&buf, buflen, tag ? "%T%c" : "%c", cap)`?

Reply:
Yeah, this was a suggestion from Alex as a way to avoid the need for flags. I released that we don't need a way to specify untagged since you can always call with `cheri_cleartag(cap)`, but you do need a way to pretend it's tagged. There's currently an edge case around null-derived capabilities that I need to think more about.

Reply:
Note that gdb can't use this as strfcap() only works for the "native" format and not for cross-debugging.  truss would use this though.

Reply:
> Some other thoughts / ideas:
> 
> * Do we want to allow field widths with zero padding for address/top/bottom/offset?  %#p permits this via its precision parameter.  You could make the width of %c act like the precision of %#p, so `%16c` would be the same as `%#.16p`.

I've added precision and padding.

> * %o for offset
> * %t for tag (0/1)
> * %l for length

I've added a bunch more options including offset, top, tag, length, and type. I'm somewhat tempted to add a variant of `%s` (type) that prints `unsealed` and `sentry` where appropriate and otherwise prints a number.

> * Should the base always be 16 or should it be selectable, e.g. `%l` (decimal) vs `%xl` (hex)?  If you go down this route, then perhaps allowing `#` to set a leading 0x prefix for non-zero values might also make sense.  While addresses are rarely things you'd want as anything other than hex, offset and length might make sense as decimal in some cases.

I can see a couple options here. We could do `x` and `d` modifiers or we could switch to making the numbers modify `x`, `X`, and `d` conversion specifications a la printf.

Reply:
*very confused why GitHub put this comment here...*

Reply:
I'm also confused by GH.  I'm not sure which of the two options I prefer, I think for length (for example) there would be three options:

1. Default to base 16 and use `d` to select base 10:
`%l` - hex
`%dl` - decimal

2. Similar to above, but default to base 10 to match printf(3) and support `#`:
`%l` - decimal
`%xl` - hex
`%#xl` - hex with prefix

3. Use the base as the base type and use the field selector as the modifier:
`%ld` - decimal
`%lx` - hex
`%#lx` - hex with prefix

I think what I find odd about 3) is that a bare `%l` no longer works and that the base feels more natural as a modifier than the mandatory "thing" to format.  Either of 1) or 2) more closely matches strftime() I think.  I might have a slight preference for 2) over 1).

Reply:
I feel like I'd end up typing a lot of `x`s with 2, but it is more natural in particular because it allows `%Xl` to mean upper-case hex.

Reply:
Yeah, I do think we will end up using `x` often for addresses, but I think for length and offset we would probably be happy to often use decimal as is typically done with `size_t` values.